File: Stripe/STPCardLoadingIndicator.m
//
//  STPCardLoadingIndicator.m
//  StripeiOS
//
//  Created by Cameron Sabol on 8/24/20.
//  Copyright © 2020 Stripe, Inc. All rights reserved.
//

#import "STPCardLoadingIndicator.h"

NS_ASSUME_NONNULL_BEGIN

static const CGFloat kCardLoadingIndicatorDiameter = 14.f;
static const CGFloat kCardLoadingInnerCircleDiameter = 10.f;
static const CFTimeInterval kLoadingAnimationSpinDuration = 0.6;

static NSString * const kLoadingAnimationIdentifier = @"STPCardLoadingIndicator.spinning";

@implementation STPCardLoadingIndicator {
    CALayer *_indicatorLayer;
}

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        self.backgroundColor = [UIColor colorWithRed:79.f/255.f green:86.f/255.f blue:107.f/255.f alpha:1.f];
        
        // Make us a circle
        CAShapeLayer *shape = [CAShapeLayer layer];
        UIBezierPath *path = [UIBezierPath bezierPathWithArcCenter:CGPointMake(0.5f*kCardLoadingIndicatorDiameter, 0.5f*kCardLoadingIndicatorDiameter)
                                                            radius:0.5f*kCardLoadingIndicatorDiameter
                                                        startAngle:0.f
                                                          endAngle:(2.f*M_PI)
                                                         clockwise:YES];
        shape.path = path.CGPath;
        self.layer.mask = shape;
        
        // Add the inner circle
        CAShapeLayer *innerCircle = [CAShapeLayer layer];
        innerCircle.anchorPoint = CGPointMake(0.5f, 0.5f);
        innerCircle.position = CGPointMake(0.5f*kCardLoadingIndicatorDiameter, 0.5f*kCardLoadingIndicatorDiameter);

        UIBezierPath *indicatorPath = [UIBezierPath bezierPathWithArcCenter:CGPointMake(0.f, 0.f)
                                                                     radius:0.5f*kCardLoadingInnerCircleDiameter
                                                                 startAngle:0.f
                                                                   endAngle:(9.f*M_PI/6.f)
                                                                  clockwise:YES];
        innerCircle.path = indicatorPath.CGPath;
        innerCircle.strokeColor = [UIColor colorWithWhite:1.f alpha:0.8f].CGColor;
        innerCircle.fillColor = [UIColor clearColor].CGColor;
        [self.layer addSublayer:innerCircle];
        _indicatorLayer = (CALayer *)innerCircle;
    }
    
    return self;
}

- (CGSize)intrinsicContentSize {
    return CGSizeMake(kCardLoadingIndicatorDiameter, kCardLoadingIndicatorDiameter);
}

- (void)didMoveToWindow {
    [super didMoveToWindow];
    [self startAnimating];
}

- (void)startAnimating {
    CABasicAnimation *spinAnimation = [CABasicAnimation animationWithKeyPath:@"transform.rotation"];
    spinAnimation.byValue = [NSNumber numberWithFloat:(float)(2.0f*M_PI)];
    spinAnimation.duration = kLoadingAnimationSpinDuration;
    spinAnimation.repeatCount = INFINITY;

    [_indicatorLayer addAnimation:spinAnimation forKey:kLoadingAnimationIdentifier];
}

- (void)stopAnimating {
    [_indicatorLayer removeAnimationForKey:kLoadingAnimationIdentifier];
}

@end

NS_ASSUME_NONNULL_END

File: StripePaymentsUI/StripePaymentsUI/Internal/UI/Views/STPCardLoadingIndicator.swift
//
//  STPCardLoadingIndicator.swift
//  StripePaymentsUI
//
//  Created by Cameron Sabol on 8/24/20.
//  Copyright © 2020 Stripe, Inc. All rights reserved.
//

import UIKit

private let kCardLoadingIndicatorDiameter: CGFloat = 14.0
private let kCardLoadingInnerCircleDiameter: CGFloat = 10.0
private let kLoadingAnimationSpinDuration: CFTimeInterval = 0.6
private let kLoadingAnimationIdentifier = "STPCardLoadingIndicator.spinning"

class STPCardLoadingIndicator: UIView {
    private var indicatorLayer: CALayer?

    override init(
        frame: CGRect
    ) {
        super.init(frame: frame)
        backgroundColor = UIColor(
            red: 79.0 / 255.0,
            green: 86.0 / 255.0,
            blue: 107.0 / 255.0,
            alpha: 1.0
        )

        // Make us a circle
        let shape = CAShapeLayer()
        let path = UIBezierPath(
            arcCenter: CGPoint(
                x: 0.5 * kCardLoadingIndicatorDiameter,
                y: 0.5 * kCardLoadingIndicatorDiameter
            ),
            radius: 0.5 * kCardLoadingIndicatorDiameter,
            startAngle: 0.0,
            endAngle: 2.0 * .pi,
            clockwise: true
        )
        shape.path = path.cgPath
        layer.mask = shape

        // Add the inner circle
        let innerCircle = CAShapeLayer()
        innerCircle.anchorPoint = CGPoint(x: 0.5, y: 0.5)
        innerCircle.position = CGPoint(
            x: 0.5 * kCardLoadingIndicatorDiameter,
            y: 0.5 * kCardLoadingIndicatorDiameter
        )

        let indicatorPath = UIBezierPath(
            arcCenter: CGPoint(x: 0.0, y: 0.0),
            radius: 0.5 * kCardLoadingInnerCircleDiameter,
            startAngle: 0.0,
            endAngle: 9.0 * .pi / 6.0,
            clockwise: true
        )
        innerCircle.path = indicatorPath.cgPath
        innerCircle.strokeColor = UIColor(white: 1.0, alpha: 0.8).cgColor
        innerCircle.fillColor = UIColor.clear.cgColor
        layer.addSublayer(innerCircle)
        indicatorLayer = innerCircle
    }

    override var intrinsicContentSize: CGSize {
        return CGSize(width: kCardLoadingIndicatorDiameter, height: kCardLoadingIndicatorDiameter)
    }

    override func systemLayoutSizeFitting(
        _ targetSize: CGSize,
        withHorizontalFittingPriority horizontalFittingPriority: UILayoutPriority,
        verticalFittingPriority: UILayoutPriority
    ) -> CGSize {
        return intrinsicContentSize
    }

    override func didMoveToWindow() {
        super.didMoveToWindow()
        startAnimating()
    }

    func startAnimating() {
        let spinAnimation = CABasicAnimation(keyPath: "transform.rotation")
        spinAnimation.byValue = NSNumber(value: Float(2.0 * .pi))
        spinAnimation.duration = kLoadingAnimationSpinDuration
        spinAnimation.repeatCount = .infinity

        indicatorLayer?.add(spinAnimation, forKey: kLoadingAnimationIdentifier)
    }

    func stopAnimating() {
        indicatorLayer?.removeAnimation(forKey: kLoadingAnimationIdentifier)
    }

    required init?(
        coder aDecoder: NSCoder
    ) {
        super.init(coder: aDecoder)
    }
}
